File: frontend/js/websocket.js
/**
 * RedStrike.AI - WebSocket Client
 */

class RedStrikeWebSocket {
    constructor() {
        this.connections = new Map();
        this.listeners = new Map();
    }

    connect(projectId) {
        if (this.connections.has(projectId)) {
            return this.connections.get(projectId);
        }

        const token = localStorage.getItem('token');
        const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
        const wsUrl = `${protocol}//${window.location.host}/ws/projects/${projectId}?token=${token}`;

        const ws = new WebSocket(wsUrl);

        ws.onopen = () => {
            console.log(`WebSocket connected for project ${projectId}`);
        };

        ws.onmessage = (event) => {
            try {
                const data = JSON.parse(event.data);
                this.handleMessage(projectId, data);
            } catch (e) {
                console.error('WebSocket message parse error:', e);
            }
        };

        ws.onclose = () => {
            console.log(`WebSocket disconnected for project ${projectId}`);
            this.connections.delete(projectId);
        };

        ws.onerror = (error) => {
            console.error('WebSocket error:', error);
        };

        this.connections.set(projectId, ws);
        return ws;
    }

    disconnect(projectId) {
        const ws = this.connections.get(projectId);
        if (ws) {
            ws.close();
            this.connections.delete(projectId);
        }
    }

    disconnectAll() {
        this.connections.forEach((ws, projectId) => {
            ws.close();
        });
        this.connections.clear();
    }

    handleMessage(projectId, data) {
        const listeners = this.listeners.get(projectId) || [];

        switch (data.type) {
            case 'scan_update':
                listeners.forEach(l => l.onScanUpdate?.(data));
                break;
            case 'new_finding':
                listeners.forEach(l => l.onNewFinding?.(data.finding));
                break;
            case 'new_endpoint':
                listeners.forEach(l => l.onNewEndpoint?.(data.endpoint));
                break;
            case 'scan_complete':
                listeners.forEach(l => l.onScanComplete?.(data.summary));
                break;
            case 'pong':
                // Heartbeat response
                break;
            default:
                console.log('Unknown message type:', data.type);
        }
    }

    subscribe(projectId, listener) {
        if (!this.listeners.has(projectId)) {
            this.listeners.set(projectId, []);
        }
        this.listeners.get(projectId).push(listener);

        // Return unsubscribe function
        return () => {
            const listeners = this.listeners.get(projectId);
            const index = listeners.indexOf(listener);
            if (index > -1) {
                listeners.splice(index, 1);
            }
        };
    }

    sendPing(projectId) {
        const ws = this.connections.get(projectId);
        if (ws && ws.readyState === WebSocket.OPEN) {
            ws.send(JSON.stringify({ type: 'ping' }));
        }
    }
}

// Global WebSocket instance
const wsClient = new RedStrikeWebSocket();

File: app/api/websocket.py
"""
RedStrike.AI - WebSocket API for Real-time Updates
"""
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from typing import Dict, Set
import json
import logging

from app.core.database import get_db
from app.core.security import decode_token
from app.models import Project

router = APIRouter(tags=["WebSocket"])
logger = logging.getLogger(__name__)


class ConnectionManager:
    """Manages WebSocket connections per project."""
    
    def __init__(self):
        # project_id -> set of websockets
        self.active_connections: Dict[int, Set[WebSocket]] = {}
    
    async def connect(self, websocket: WebSocket, project_id: int):
        """Accept and register a new connection."""
        await websocket.accept()
        
        if project_id not in self.active_connections:
            self.active_connections[project_id] = set()
        
        self.active_connections[project_id].add(websocket)
        logger.info(f"Client connected to project {project_id}")
    
    def disconnect(self, websocket: WebSocket, project_id: int):
        """Remove a connection."""
        if project_id in self.active_connections:
            self.active_connections[project_id].discard(websocket)
            if not self.active_connections[project_id]:
                del self.active_connections[project_id]
        logger.info(f"Client disconnected from project {project_id}")
    
    async def send_to_project(self, project_id: int, message: dict):
        """Send message to all clients watching a project."""
        if project_id in self.active_connections:
            dead_connections = set()
            
            for websocket in self.active_connections[project_id]:
                try:
                    await websocket.send_json(message)
                except Exception as e:
                    logger.error(f"Failed to send message: {e}")
                    dead_connections.add(websocket)
            
            # Clean up dead connections
            for ws in dead_connections:
                self.active_connections[project_id].discard(ws)
    
    async def broadcast_scan_update(
        self,
        project_id: int,
        phase: str,
        status: str,
        message: str,
        data: dict = None,
    ):
        """Broadcast scan progress update."""
        await self.send_to_project(project_id, {
            "type": "scan_update",
            "phase": phase,
            "status": status,
            "message": message,
            "data": data or {},
        })
    
    async def broadcast_finding(self, project_id: int, finding: dict):
        """Broadcast new finding."""
        await self.send_to_project(project_id, {
            "type": "new_finding",
            "finding": finding,
        })
    
    async def broadcast_endpoint(self, project_id: int, endpoint: dict):
        """Broadcast new endpoint discovered."""
        await self.send_to_project(project_id, {
            "type": "new_endpoint",
            "endpoint": endpoint,
        })
    
    async def broadcast_scan_complete(self, project_id: int, summary: dict):
        """Broadcast scan completion."""
        await self.send_to_project(project_id, {
            "type": "scan_complete",
            "summary": summary,
        })


# Global connection manager
manager = ConnectionManager()


@router.websocket("/ws/projects/{project_id}")
async def websocket_endpoint(
    websocket: WebSocket,
    project_id: int,
):
    """WebSocket endpoint for real-time project updates."""
    
    # Authenticate via query param token
    token = websocket.query_params.get("token")
    if not token:
        await websocket.close(code=4001, reason="Authentication required")
        return
    
    try:
        payload = decode_token(token)
        user_id = int(payload.get("sub"))
    except Exception as e:
        await websocket.close(code=4001, reason="Invalid token")
        return
    
    # Connect
    await manager.connect(websocket, project_id)
    
    try:
        while True:
            # Keep connection alive, handle any client messages
            data = await websocket.receive_text()
            
            try:
                message = json.loads(data)
                
                # Handle client commands
                if message.get("type") == "ping":
                    await websocket.send_json({"type": "pong"})
                
            except json.JSONDecodeError:
                pass
                
    except WebSocketDisconnect:
        manager.disconnect(websocket, project_id)


# Export manager for use in other modules
def get_connection_manager() -> ConnectionManager:
    return manager
